flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -510,831 +505,92 @@ public WriteOptions getWriteOptions() {
-					RegisteredStateMetaInfoBase stateMetaInfo =
-						RegisteredStateMetaInfoBase.fromMetaInfoSnapshot(restoredMetaInfo);
-					registeredColumn = new Tuple2<>(columnFamily, stateMetaInfo);
-					rocksDBKeyedStateBackend.kvStateInformation.put(restoredMetaInfo.getName(), registeredColumn);
-
-				} else {
-					// TODO with eager state registration in place, check here for serializer migration strategies
-				}
-				currentStateHandleKVStateColumnFamilies.add(registeredColumn.f0);
-			}
-		}
-
-		/**
-		 * Restore the KV-state / ColumnFamily data for all key-groups referenced by the current state handle.
-		 */
-		private void restoreKVStateData() throws IOException, RocksDBException {
-			//for all key-groups in the current state handle...
-			try (RocksDBWriteBatchWrapper writeBatchWrapper = new RocksDBWriteBatchWrapper(rocksDBKeyedStateBackend.db)) {
-				for (Tuple2<Integer, Long> keyGroupOffset : currentKeyGroupsStateHandle.getGroupRangeOffsets()) {
-					int keyGroup = keyGroupOffset.f0;
-
-					// Check that restored key groups all belong to the backend
-					Preconditions.checkState(rocksDBKeyedStateBackend.getKeyGroupRange().contains(keyGroup),
-						"The key group must belong to the backend");
-
-					long offset = keyGroupOffset.f1;
-					//not empty key-group?
-					if (0L != offset) {
-						currentStateHandleInStream.seek(offset);
-						try (InputStream compressedKgIn = keygroupStreamCompressionDecorator.decorateWithCompression(currentStateHandleInStream)) {
-							DataInputViewStreamWrapper compressedKgInputView = new DataInputViewStreamWrapper(compressedKgIn);
-							//TODO this could be aware of keyGroupPrefixBytes and write only one byte if possible
-							int kvStateId = compressedKgInputView.readShort();
-							ColumnFamilyHandle handle = currentStateHandleKVStateColumnFamilies.get(kvStateId);
-							//insert all k/v pairs into DB
-							boolean keyGroupHasMoreKeys = true;
-							while (keyGroupHasMoreKeys) {
-								byte[] key = BytePrimitiveArraySerializer.INSTANCE.deserialize(compressedKgInputView);
-								byte[] value = BytePrimitiveArraySerializer.INSTANCE.deserialize(compressedKgInputView);
-								if (hasMetaDataFollowsFlag(key)) {
-									//clear the signal bit in the key to make it ready for insertion again
-									clearMetaDataFollowsFlag(key);
-									writeBatchWrapper.put(handle, key, value);
-									//TODO this could be aware of keyGroupPrefixBytes and write only one byte if possible
-									kvStateId = END_OF_KEY_GROUP_MARK
-										& compressedKgInputView.readShort();
-									if (END_OF_KEY_GROUP_MARK == kvStateId) {
-										keyGroupHasMoreKeys = false;
-									} else {
-										handle = currentStateHandleKVStateColumnFamilies.get(kvStateId);
-									}
-								} else {
-									writeBatchWrapper.put(handle, key, value);
-								}
-							}
-						}
-					}
-				}
-			}
+			RocksDBRestoreOperation<K> restoreOperation = getRocksDBRestoreOperation(
```

Comment:
Yes required by `StateBackendTestContext.restoreSnapshot`, and yes that's the reason `restoreOperation.restore` requires a `db` parameter. I'm not sure whether we need to reserve a standalone restore function for future usage, but will try to get rid of it with your confirmation.

Reply:
As mentioned in my previous comment, I think we don't need a standalone `restore` in that `StateBackendTestContext`, the existence of that function was only because of problems that go away now with your refactoring. You can also see that create and restore are always used together. I confirm that we can merge them.

Reply:
For the current codes, yes I agree, while my questions is actually whether we need to reserve it for future usage, although I don't know what exactly the requirement will be. Or maybe we could add it back whenever the need emerges, won't be hard anyway (smile).

Reply:
I would take care of the current code, I don't think there are many new concerns added in the future. The separation of create and restore was never something wanted, just a consequence of the changed order introduced with incremental snapshots, that the db object is only completed after the restore method was called, which also coupled the calls of create and restore together all the time. So there is no regression to be afraid about.